### PR TITLE
refactor: consolidate HACS read tools from 4 to 2

### LINKED ISSUE #833: [FEATURE] Consolidate 4 read-only HACS tools into a single tool
### What would you like to see?

The 4 read-only HACS tools (`ha_hacs_info`, `ha_hacs_list_installed`, `ha_hacs_search`, `ha_hacs_repository_info`) could be consolidated into a single tool, reducing cognitive load for AI agents.

Currently these tools all query HACS via WebSocket:
- `ha_hacs_info()` — returns HACS version/status/categories (`hacs/info`)
- `ha_hacs_list_installed(category?)` — lists installed repos (`hacs/repositories/list` filtered to installed)
- `ha_hacs_search(query, category?, ...)` — searches all repos (`hacs/repositories/list` with text matching)
- `ha_hacs_repository_info(repository_id)` — gets detailed repo info + README (`hacs/repository/info`)

`ha_hacs_list_installed` and `ha_hacs_search` already call the same WebSocket command (`hacs/repositories/list`) — one filters to installed repos, the other does text matching. These are natural merge candidates.

**Proposed combined tool** (name TBD, e.g. `ha_get_hacs_repository`):

| Params provided | Behavior | Replaces |
|---|---|---|
| Nothing | Return HACS system info (version, categories, stage) | `ha_hacs_info` |
| `installed_only=True` | List installed repos | `ha_hacs_list_installed` |
| `query="mushroom"` | Search repos by keyword | `ha_hacs_search` |
| `repository_id="441028036"` | Get detailed repo info + README | `ha_hacs_repository_info` |

The 2 write tools (`ha_hacs_add_repository`, `ha_hacs_download`) would remain separate.

Additionally, the current HACS tool names don't follow the repo's `ha_<verb>_<noun>` convention (e.g. `ha_hacs_info` is missing a verb). The consolidated tool would fix this.

### Why do you need this?

Per the project's own design principles (AGENTS.md): "When a tool's functionality is fully covered by another tool, remove the redundant tool. Fewer tools reduce cognitive load for AI agents and improve decision-making."

An LLM currently has to decide between 4 different HACS read tools that overlap significantly. A single tool with clear mode selection is simpler to reason about. Two of the tools already hit the exact same API endpoint.

### Additional context

This is a proposal — I'm open to discussion on the approach, naming, parameter design, or whether to keep any of the tools separate. Happy to implement if there's interest, or drop it entirely if the maintainers prefer the current structure.

### PR DESCRIPTION
**Pull request by kingpanther13**
## What does this PR do?

Consolidates the 4 read-only HACS tools down to 2, based on the discussion in #833:

1. **Merge `ha_hacs_list_installed` into `ha_hacs_search`** — adds `installed_only` parameter. Both tools already called the same `hacs/repositories/list` WebSocket endpoint. `ha_hacs_search(installed_only=True)` replaces `ha_hacs_list_installed()`.

2. **Remove `ha_hacs_info` entirely** — every field it returned is available via `ha_get_system_health` (version, stage, lovelace_mode). HACS availability is already checked internally by `_assert_hacs_available()` in every HACS tool. The `categories` field isn't actionable — valid values are already hardcoded as `Literal` types in the tool parameter schemas.

**Result:** 4 read-only HACS tools → 2 (`ha_hacs_search`, `ha_hacs_repository_info`). The 2 write tools (`ha_hacs_add_repository`, `ha_hacs_download`) are unchanged.

### ha_hacs_search changes

- `query` is now optional (default `""`) — empty query with `installed_only=True` lists all installed repos
- New `installed_only` param (default `False`) filters to installed repos only
- When `installed_only=True`, response includes `pending_update` and `domain` fields per repo
- When listing (no query), results sorted alphabetically by name instead of by score

## Type of change
- [ ] 🐛 Bug fix
- [ ] ✨ New feature
- [ ] 📚 Documentation
- [x] 🔧 Maintenance/refactor
- [ ] 💥 Breaking change

## Testing
- [ ] I have tested these changes with a LLM agent
- [x] All automated tests pass (`uv run pytest`)
- [x] Code follows style guidelines (`uv run ruff check`)

**Test updates:**
- Removed `TestHacsInfo` class
- Renamed `TestHacsListInstalled` → `TestHacsSearchInstalled`, updated to use `ha_hacs_search(installed_only=True)`
- Updated all HACS availability checks from `ha_hacs_info` to `ha_hacs_search`

## Checklist
- [x] I have updated documentation if needed

Closes #833

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/ha_mcp/tools/tools_hacs.py [hunk 1]
```diff
@@ -262,56 +111,54 @@ async def ha_hacs_search(
...
-
-        Returns:
-            Search results from HACS store or error if HACS is not available.
         """
         try:
-            # Coerce max_results and offset to int
+            # Coerce parameters
+            installed_only_bool = (
+                coerce_bool_param(installed_only, "installed_only", default=False)
+                or False
```
Comment (sergeykad) [inline src/ha_mcp/tools/tools_hacs.py]: **1. `or False` is dead code**

`coerce_bool_param(..., default=False)` never returns `None` when a non-None default is provided — same reason the `or 10` / `or 0` were correctly removed from the int coercions below. This `or False` should be removed for consistency:

```python
installed_only_bool = coerce_bool_param(installed_only, "installed_only", default=False)
```

File: src/ha_mcp/tools/tools_hacs.py [hunk 2]
```diff
@@ -53,200 +54,48 @@ async def _assert_hacs_available() -> None:
...
+        annotations={
+            "idempotentHint": True,
+            "readOnlyHint": True,
+            "title": "Search HACS Store",
+        },
+    )
     @log_tool_usage
     async def ha_hacs_search(
-        query: str,
+        query: str = "",
```
Comment (sergeykad) [inline src/ha_mcp/tools/tools_hacs.py:99]: **2. Empty query + `installed_only=False` returns all repos with no validation**

With `query` now defaulting to `""`, calling `ha_hacs_search()` with no arguments silently returns all HACS store repos sorted alphabetically — the response looks successful so an LLM won't know it did something unintended. Either validate that at least one of `query`/`installed_only` is provided, or document this as intentional "browse all" behavior in the docstring.

File: tests/src/e2e/workflows/hacs/test_hacs.py [hunk 3]
```diff
@@ -91,121 +107,43 @@ def is_hacs_unavailable(data: dict) -> tuple[bool, str]:
...
-        logger.info("HACS info response structure test passed")
-
-
-@pytest.mark.hacs
-class TestHacsListInstalled:
-    """Test HACS list installed repositories functionality."""
+class TestHacsSearchInstalled:
+    """Test HACS search with installed_only filter functionality."""
 
     async def test_list_all_installed(self, mcp_client):
```
Comment (sergeykad) [inline tests/src/e2e/workflows/hacs/test_hacs.py:113]: **3. Missing test: `installed_only=True` with non-empty query**

The code path where `installed_only=True` AND `query` is non-empty (filter to installed, then score by relevance) is never tested. If the `installed_only` filter were accidentally placed after the scoring block, non-installed repos matching the query would leak through. Consider adding a test like:

```python
async def test_search_installed_with_query(self, mcp_client):
    data = await safe_hacs_call(mcp_client, "ha_hacs_search", {"query": "hacs", "installed_only": True})
    # assert all results have installed=True
```

File: README.md [hunk 4]
```diff
@@ -146,7 +146,7 @@ Spend less time configuring, more time enjoying your smart home.
 | **Entity Registry** | `ha_get_entity_exposure`, `ha_get_entity`, `ha_set_entity` |
 | **Files** | `ha_delete_file`, `ha_list_files`, `ha_read_file`, `ha_write_file` |
 | **Groups** | `ha_config_list_groups`, `ha_config_remove_group`, `ha_config_set_group` |
-| **HACS** | `ha_hacs_add_repository`, `ha_hacs_download`, `ha_hacs_info`, `ha_hacs_list_installed`, `ha_hacs_repository_info`, `ha_hacs_search` |
+| **HACS** | `ha_hacs_add_repository`, `ha_hacs_download`, `ha_hacs_repository_info`, `ha_hacs_search` |
```
Comment (sergeykad) [inline README.md:149]: **4. Stale references to removed tools**

`homeassistant-addon/DOCS.md:379-380` still lists `ha_hacs_info` and `ha_hacs_list_installed`, and `site/src/data/tools.json` still has entries for both. Run `python scripts/extract_tools.py` to regenerate tools.json and update DOCS.md to match.

File: src/ha_mcp/tools/tools_hacs.py [hunk 5]
```diff
@@ -450,66 +337,83 @@ async def ha_hacs_repository_info(repository_id: str) -> dict[str, Any]:
...
-                            "suggestions": [
-                                "Use ha_hacs_search() to find the repository",
-                                "Check the repository name is correct (case-insensitive)",
-                                "The repository may need to be added to HACS first",
-                            ],
-                        })
+                        return await add_timezone_metadata(
+                            client,
+                            {
+                                "success": False,
```
Comment (sergeykad) [inline src/ha_mcp/tools/tools_hacs.py]: **5. `ha_hacs_repository_info` returns `success: False` instead of raising `ToolError`**

Per project guidelines: "All tool-level failures must raise `ToolError`." This "repository not found" path returns `{"success": False}` with `isError=false` in MCP, so an LLM may not recognize it as an error. Should use `raise_tool_error(create_error_response(...))`. Same pattern at line 469 (`ha_hacs_add_repository` invalid format) and line 611 (`ha_hacs_download` not found).

File: src/ha_mcp/tools/tools_hacs.py [hunk 6]
```diff
@@ -40,6 +40,7 @@ async def _is_hacs_available() -> bool:
     their own exception_to_structured_error blocks.
     """
     from ..client.websocket_client import get_websocket_client
+
     ws_client = await get_websocket_client()
```
Comment (sergeykad) [inline src/ha_mcp/tools/tools_hacs.py:48]: **6. `_is_hacs_available` conflates connection errors with HACS absence**

If `hacs/info` returns `{"success": false, "error": ...}` (HACS installed but broken), `_is_hacs_available` returns `False` and the caller tells the user "HACS is not installed" — wrong diagnosis. With this PR removing `ha_hacs_info`, `_assert_hacs_available` becomes the sole HACS gatekeeper, making this more impactful.

Should inspect the response error to distinguish "unknown command" (not installed) from other failures, and `raise_tool_error` with the appropriate error code (e.g. `COMPONENT_NOT_INSTALLED` vs `SERVICE_CALL_FAILED`) so the LLM gets an actionable `isError=true` response with the correct diagnosis.

File: tests/src/e2e/workflows/hacs/test_hacs.py [hunk 7]
```diff
@@ -91,121 +107,43 @@ def is_hacs_unavailable(data: dict) -> tuple[bool, str]:
...
-        assert "total_installed" in data, "Response should include total_installed"
-        assert "repositories" in data, "Response should include repositories list"
+        assert "total_matches" in data, "Response should include total_matches"
+        assert "results" in data, "Response should include results list"
         assert "category_filter" in data, "Response should include category_filter"
 
-        total = data["total_installed"]
-        repos = data["repositories"]
+        total = data["total_matches"]
+        repos = data["results"]
```
Comment (sergeykad) [inline tests/src/e2e/workflows/hacs/test_hacs.py:140]: **7. Response doesn't assert `installed_only` field**

`test_list_all_installed` checks `total_matches`, `results`, and `category_filter` but doesn't assert that the new `installed_only` field is present and `True` in the response. Minor but documents the API contract.

File: src/ha_mcp/tools/tools_hacs.py (hunk 1 above)
Comment (kingpanther13) [reply src/ha_mcp/tools/tools_hacs.py]: Fixed in 6814293 — removed the dead `or False`.

File: src/ha_mcp/tools/tools_hacs.py (hunk 2 above)
Comment (kingpanther13) [reply src/ha_mcp/tools/tools_hacs.py:99]: This is intentional "browse all" behavior. Calling ha_hacs_search() with no args returns all HACS store repos, same as browsing the HACS store with no filters. The docstring documents the search vs listing modes. An LLM calling with no args gets a paginated, sorted list — not an error condition.

File: tests/src/e2e/workflows/hacs/test_hacs.py (hunk 3 above)
Comment (kingpanther13) [reply tests/src/e2e/workflows/hacs/test_hacs.py:113]: Good catch — added `test_search_installed_with_query` in 6814293. Exercises the installed_only=True + non-empty query code path and asserts all returned repos have installed=True.

File: README.md (hunk 4 above)
Comment (kingpanther13) [reply README.md:149]: tools.json, README.md, and DOCS.md are all auto-regenerated by the sync-tool-docs.yml workflow on merge (added in PRs #839 and #849). The workflow runs `python scripts/extract_tools.py` and commits the result. No manual updates needed — the stale references will be cleaned up automatically when this PR merges.

File: src/ha_mcp/tools/tools_hacs.py (hunk 5 above)
Comment (kingpanther13) [reply src/ha_mcp/tools/tools_hacs.py]: Good catch — fixed all three locations in 6814293. ha_hacs_repository_info and ha_hacs_download now raise ToolError with RESOURCE_NOT_FOUND for "not found" cases, and ha_hacs_add_repository raises with VALIDATION_INVALID_PARAMETER for bad format. These were pre-existing patterns carried over from the old code.

File: src/ha_mcp/tools/tools_hacs.py (hunk 6 above)
Comment (kingpanther13) [reply src/ha_mcp/tools/tools_hacs.py:48]: Fixed in 6814293 — _assert_hacs_available now inspects the error response from hacs/info. Distinguishes "unknown_command" (HACS not installed) from other failures (HACS installed but broken) with different error messages and suggestions. Also removed the separate _is_hacs_available helper since _assert_hacs_available is now the sole gatekeeper.

File: tests/src/e2e/workflows/hacs/test_hacs.py (hunk 7 above)
Comment (kingpanther13) [reply tests/src/e2e/workflows/hacs/test_hacs.py:140]: Added in 6814293 — `test_list_all_installed` now asserts `data["installed_only"] is True`.

File: src/ha_mcp/tools/tools_hacs.py (hunk 2 above)
Comment (sergeykad) [reply src/ha_mcp/tools/tools_hacs.py:99]: This was resolved without a fix or explanation. The issue still stands — calling `ha_hacs_search()` with no arguments returns all HACS store repos silently. Please either add validation or document this as intentional.

File: README.md (hunk 4 above)
Comment (sergeykad) [reply README.md:149]: This was resolved without a fix. `homeassistant-addon/DOCS.md` and `site/src/data/tools.json` still reference `ha_hacs_info` and `ha_hacs_list_installed`. Please update DOCS.md and run `python scripts/extract_tools.py` to regenerate tools.json.

Comment (kingpanther13) [reply README.md:149]: @sergeykad there is no fix to resolve here. As my previous comment states, the workflow runs it automatically now per #839 and #849.

File: src/ha_mcp/tools/tools_hacs.py (hunk 2 above)
Comment (kingpanther13) [reply src/ha_mcp/tools/tools_hacs.py:99]: @sergeykad Explanation is right there in my last comment, this is intentional.

Comment (sergeykad) [reply src/ha_mcp/tools/tools_hacs.py:99]: The "browse all" behavior is fine if documented. Currently the docstring only describes "Search mode" and "Installed mode" — it does not mention that an empty query without `installed_only` returns all repos. Please add a note, e.g. a third mode or a line under the `query` param description.

Comment (kingpanther13) [reply src/ha_mcp/tools/tools_hacs.py:99]: Fair point — added explicit "Browse mode" documentation to the docstring in d8b7d19. The no-query behavior is now called out as a distinct mode alongside Search and Installed modes.
